humility-core/src/hubris.rs

```diff
@@ -570,9 +580,60 @@ impl HubrisArchive {
             qualified_variables: MultiMap::new(),
             unions: HashMap::new(),
             definitions: MultiMap::new(),
+            namespaces: Vec::new(),
         })
     }
 
+    //
+    // For a namespace identifer, compose and return the entire namespace
+    // vector as strings.
+    //
+    fn namespace(&self, namespace: Option<usize>) -> Result<Vec<&String>> {
+        let mut rval = vec![];
+        let mut current = namespace;
+
+        while let Some(namespace) = current {
+            if namespace >= self.namespaces.len() {
+                bail!("namespace id {namespace} is invalid");
+            }
+
+            let (name, parent) = &self.namespaces[namespace];
+            rval.push(name);
+            current = *parent;
+        }
+
+        rval.reverse();
+
+        Ok(rval)
+    }
+
+    //
+    // A convenience routine to take a namespace identifier and a name,
+    // and return the entire, ::-delimited name.
+    //
+    fn namespaced_name(
+        &self,
+        namespace: Option<usize>,
+        name: &String,
+    ) -> Result<Option<String>> {
+        let mut n = self.namespace(namespace)?;
+
+        if !n.is_empty() {
+            n.push(name);
+
+            let full = n
+                .iter()
+                .copied()
+                .map(String::as_str)
+                .collect::<Vec<_>>()
+                .join("::");
```

Comment:
If you change `fn namespace` to return a `Result<Vec<&str>>`, this can simply become
```suggestion
            let full = n.join("::");
```